Add flag to follow symbolic links when crawling files

viddo opened the pull request:
**Summary**
At the moment symbolic links aren't followed, because (jest-haste-map) crawler only finds _regular files_. Modified the `nativeFind` function to also yield files of type `l`, i.e. symlinks.

This resolves #1477. As suggested by [@cpojerin](https://github.com/facebook/jest/issues/1477#issuecomment-259253916) there, I've added support behind a flag that also disabled watchman, for this to work.

Note that this change don't add support for win32 platform, since that uses a [different code path/find method](https://github.com/facebook/jest/blob/24a61bb1aa310d92208833c33fa6fb8b1cfe4d2c/packages/jest-haste-map/src/crawlers/node.js#L145-L146). I don't use Windows myself, and don't know how symlinks actually work there to be honest. My rationale here is that this change set adds value that is better than no support at all. If it would make sense win32 code path could always be added later, I figure.

**Test plan**
- Updating existing tests
- Add test case for new flag modifying the find command to include symlinks
- Reproducible example case can be found at [viddo/jest-symlinks-probing](https://github.com/viddo/jest-symlinks-probing). See the README there for instructions how to reproduce and the example that validates this change
  - I'm happy to convert this into a integration test, but not entirely sure what the expectation is here. Any pointers? I'd like to get some early feedback before I spend more time on this PR.

Thanks for reading!

SimenB commented:
>My rationale here is that this change set adds value that is better than no support at all.

Isn't there a risk that the same test suite behaves differently on differently platforms because of symbolic links?

viddo commented:
> Isn't there a risk that the same test suite behaves differently on differently platforms because of symbolic links?

I'd say it depends on the use-case. I guess you're right, that there _could_ be a risk of that for usages where both linux and windows are supported. On the other hand, do symlinks created on linux env work on a windows env? From what I can read up the answer seems to be _it depends_, too, on Windows version and how the symlink is done.

I'd imagine that existing usages of jest where both platforms are used don't have this issue, since symlinks aren't currently supported.

What do you think or have in mind?

SimenB commented:
My thinking was that _if_ this is implemented, and some mac/linux user adds a test requiring symlinks, pushes it, and the test suite is broken for all windows developers

viddo commented:
By default nothing would break. Even if using the new flag the test requiring symlinks wouldn't be run on a windows platform.

If your question (_"My thinking was that if this is implemented, …"_) refers to implementing support for Windows, I'd again say that depends on the implementation and how symlinks would be resolved across platforms.

SimenB commented:
So a windows developer won't even know that they broke a test before seeing it break on CI, and have no way of reproducing locally bar installing a new OS?

viddo commented:
Correct that a windows developer wouldn't know until the CI breaks in that scenario. However, I'd not expect a developer to install a new OS as goto solution, but investigate the CI result and be questioning why this feature was enabled and a symlink was used for a project that also have windows developers (supporting cross-platform?).

I interpret your questions as that you think the the PR and/or rationale is the wrong approach. That's fine. Rather than talking about a hypothetical case back and forth I'd prefer to have a dialogue about what change you're really asking for and/or what the expectations are?

SimenB commented:
All I'm saying is that introducing a feature that purposefully does not work on a platform supported by the software (I understand it might be impossible to implement) seems like a wrong decision. It's behind a flag, so maybe no big deal? I don't have strong (if any) feelings against this, as long as cross-platform support is not treated like "meh".

>questioning why this feature was enabled and a symlink was used for a project that also have windows developers

Don't forget that half of the developers that use nodejs are on windows<sup>1</sup> - it's not a niche group.


<sup>1 It was last year at least, quick googling didn't give any recent numbers</sup>

cpojer commented:
I have some initial thoughts and will do a more thorough review later:

* The second piece of code that uses Node's fs APIs instead of find also needs to be supported.
* The name of this option is too simple. If we intent to merge this, it needs to say loud and clear that this is not really officially supported.

Overall I almost feel like it would be better if we made crawlers pluggable – that way we can merge a PR in Jest that allows you to specify your own crawler, and you can release your own package that crawls stuff in any way you like, from any source you like.

viddo commented:
I don't think cross-platform support should be treated as a "meh" neither, sorry if it came across like that. :/ I haven't developed on/for windows platform for many years (except IE), so I simply don't feel confident about implementing support for it myself.

> - The second piece of code that uses Node's fs APIs instead of find also needs to be supported.

Ok, I'll look into this one.

From what I understand the "correct" way to have this covered in the test suite would be to add a folder under [integration_tests/symlinks](https://github.com/facebook/jest/tree/be6952dd9a4f4646a00aa0467f07b6fed9c4a12f/integration_tests), right?

> - The name of this option is too simple. If we intent to merge this, it needs to say loud and clear that this is not really officially supported.

What flag name and description would you like to have to make this "loud and clear" to be more specific?

> Overall I almost feel like it would be better if we made crawlers pluggable – that way we can merge a PR in Jest that allows you to specify your own crawler, and you can release your own package that crawls stuff in any way you like, from any source you like.

This is an interesting idea. It feels like a potential "feature creep", though. How much effort and what are the things would have to be changed to support this? I'm not too familiar with the code base just yet to be able to tell myself.

Thanks for the feedback both of you! 🙇

joearasin commented:
Just as a point of "why do this" -- I think getting something like this merged in would unstick https://github.com/facebook/metro-bundler/issues/1

viddo commented:
Implemented support in the fs API and added a integration test to have things covered. Unfortunately it turns out this doesn't work on win32 either way, as the result from Appveyor CI shows. Managed to do some troubleshooting on a windows system (v8.1), a linux symlink isn't identified a symbolic link there:
```js
// e.g. ln -s ../other.js symlink.js
fs = require('fs')
stat = fs.lstatSync('symlink.js')
stat.isSymbolicLink() // => false on windows
```

Found these resources that contain more details on the problems. Handling symlinks properly would require something more sophisticated, if it's even possible in all scenarios:
- https://blogs.windows.com/buildingapps/2016/12/02/symlinks-windows-10/#eAqgbszZOtlJRhJ0.97
- https://github.com/git-for-windows/git/wiki/Symbolic-Links

Closing this PR. What @cpojer suggested with making the crawlers pluggable sounds reasonable, but that would be a different PR.
